### PR TITLE
[feature] Add Ascend NPU Support While Preserving GPU Compatibility

### PR DESCRIPTION
**Pull request by sunny-infra**
# Add Ascend NPU Support While Preserving GPU Compatibility

## Summary

This PR extends DeepSpec to support Huawei Ascend NPUs (910B and above) as a first-class backend, alongside the existing CUDA backend. All changes are fully backward-compatible — GPU users see no behavior change. A detailed design rationale is documented in [`RFC: Add Ascend NPU Support for DeepSpec`](https://github.com/deepseek-ai/DeepSpec/issues/6).

## Motivation

The README currently states DeepSpec targets CUDA-only environments. This limits adoption on Ascend NPU clusters, which are widely deployed in China and increasingly used for LLM training. The goal of this PR is to make the entire pipeline — **target cache generation → DSpark training → speculative-decoding evaluation** — runnable on Ascend NPUs without forking the codebase, while keeping the GPU path untouched.

## Key Changes

All NPU adaptations follow a **detect-and-fallback** pattern: at runtime we check `torch.npu.is_available()` and branch; on GPU the original code path is taken verbatim.

| Area | File | Change |
|---|---|---|
| Distributed init | `deepspec/utils/distributed.py` | `init_dist` selects `hccl` backend and `torch.npu` device on NPU; `nccl`/`cuda` on GPU |
| Seeding | `deepspec/utils/__init__.py` | `seed_all` seeds NPU RNG when available |
| Data loading | `deepspec/data/cuda_prefetcher.py` | Stream module selected dynamically (`torch.npu` vs `torch.cuda`) |
| FSDP | `deepspec/trainer/base_trainer.py` | `device_mesh` uses `"npu"` or `"cuda"`; added per-epoch checkpoint saving (orthogonal improvement) |
| Checkpointing | `deepspec/trainer/ckpt_manager.py` | RNG state save/load branches by backend; tolerates legacy checkpoints without RNG state |
| Attention mask | `deepspec/modeling/dspark/common.py` | `create_dspark_attention_mask` gains `use_block_mask` flag (default `True`); when `False` produces a 4D additive mask for the eager path |
| Attention impl | `deepspec/modeling/dspark/qwen3/config.py` | `TRAIN_ATTN_IMPLEMENTATION` selected at runtime: `eager` on NPU, `flex_attention` on GPU |
| Modeling dispatch | `deepspec/modeling/dspark/qwen3/modeling.py` | Mask type chosen by `_attn_implementation` |
| Train entry | `train.py` | `spawn` nprocs from `_device_count()` (NPU-aware) |
| Eval entry | `eval.py` | `_visible_device_count()` reads `ASCEND_RT_VISIBLE_DEVICES`/`CUDA_VISIBLE_DEVICES`; adds `--tasks` flag |
| Cache generation | `scripts/data/prepare_target_cache.py` | NPU device detection and empty-cache helper |
| Eval all-reduce | `deepspec/eval/dspark/confidence_head.py` | `float64` → `float32` cast only on NPU (HCCL limitation); GPU keeps `float64` |

### New updates added on 2026-06-30

1. New Centralized Device Abstraction Layer `deepspec/utils/device.py` ! All device-related helper functions are unified in this module to avoid scattered implementations like `_is_npu_available()` across multiple entry files:

| Function | Purpose |
|---|---|
| `is_npu_available()` | Runtime detection of NPU availability |
| `device_type()` | Returns `"npu"` or `"cuda"` |
| `accelerator_module()` | Returns `torch.npu` or `torch.cuda` |
| `accelerator_backend()` | Returns `"hccl"` or `"nccl"` |
| `device_count()` / `set_device()` / `make_device()` | Device count and binding |
| `manual_seed_all()` / `get_rng_state()` / `set_rng_state()` | RNG management |
| `empty_cache()` | Memory reclamation |

**Explicit Backend Override**: Introduces the `DEEPSPEC_DEVICE` environment variable, allowing users to force `cuda` or `npu` and skip auto-detection. By default (when unset), the auto-detection path is used.

2. Added support for the SDPA backend and eliminated potential side effects introduced by eager execution mode.

3. Addressed other review comments including misaligned debug logs with the original implementation.

## Backward Compatibility

Every modification is gated by a device check, so on CUDA:

- `init_dist` still uses `nccl` + `torch.cuda`
- `seed_all` still seeds `torch.cuda`
- `CUDAPrefetcher` still uses `torch.cuda.Stream`
- FSDP `device_mesh` still uses `"cuda"`
- `create_dspark_attention_mask` defaults to `use_block_mask=True` → original `create_block_mask` path
- `TRAIN_ATTN_IMPLEMENTATION` resolves to `"flex_attention"` → original path
- `confidence_head` all-reduce keeps `float64`
- `eval.py` `_visible_device_count` falls back to `CUDA_VISIBLE_DEVICES`, behavior identical to the previous `_device_count`

No GPU code path is altered. The `--tasks` flag is optional with default `None` (run all tasks), preserving the original behavior.

## Validation

End-to-end validated on Ascend 910B and 910C (single node, 2–16 cards):

1. **Target cache generation** — 10k samples of Qwen3-8B hidden states produced (≈217 GB, verified token count)
2. **DSpark training** — 3 epochs on the 10k cache completed without OOM (with `num_anchors=256`)
3. **Evaluation** — GSM8K speculative-decoding eval ran to completion on 4 cards, producing acceptance-length / verify-rate / confidence-head metrics

GPU regression testing was not performed in this environment (no CUDA available), but every diff is structurally a no-op on CUDA per the gating described above.

## Environment

Ascend NPU users set (CUDA users do nothing):

```bash
export ASCEND_RT_VISIBLE_DEVICES=0,1,2,3,4        # equivalent to CUDA_VISIBLE_DEVICES
export PYTORCH_NPU_ALLOC_CONF=expandable_segments:True   # mitigates fragmentation
export RANK=0 WORLD_SIZE=1 MASTER_ADDR=127.0.0.1 MASTER_PORT=29500  # single-node
```

`RANK`/`WORLD_SIZE` denote **node** rank/size (single-node ⇒ always `0`/`1`); per-card rank is derived internally from `device_count()`.

## RFC

Design rationale, alternatives considered, and unresolved questions are documented in [`RFC: Add Ascend NPU Support for DeepSpec`](https://github.com/deepseek-ai/DeepSpec/issues/6). Reviewer feedback on the open questions in Section 10 is welcome.

## Contributors

| GitHub ID | Main Contributions | Details |
|-----------|---------|---------|
| [@sunny-infra ](https://github.com/[sunny-infra](https://github.com/sunny-infra)) | End-to-end Porting & Architecture | Completed end-to-end adaptation for hidden state generation, training and evaluation, drafted supporting documents and RFC, and finished full pipeline validation based on the Qwen3 model. |
| [@hazelduan ](https://github.com/hazelduan) | Refactoring & Optimization | Refactored and optimized the device abstraction layer to improve code maintainability, improved training workflows, addressed all review feedback, and completed validation with the Gemma 4 model. |
| [@wangmhua ](https://github.com/[wangmhua](https://github.com/wangmhua)) | GPU Verified | Validated all functionalities covered in this PR on GPUs to verify the compatibility of code changes. |

## Checklist

- [x] NPU path validated end-to-end (cache → train → eval)
- [x] GPU path structurally unchanged (device-gated branches)
- [x] No new hard dependencies — `torch_npu` is imported lazily inside `try/except`
- [x] RFC document included
- [x] No secrets, keys, or credentials in the diff
- [x] Commit messages follow Conventional Commits style

### CODE REVIEW

File: eval.py [hunk 1]
```diff
@@ -16,7 +18,7 @@
 }
 
 TASKS = [
-    ("gsm8k", 500),
+    ("gsm8k", 10),
```
Comment (huangxinV587) [inline eval.py]: 调试残留

File: scripts/data/prepare_target_cache.py [hunk 2]
```diff
@@ -254,7 +275,7 @@ def main(local_rank: int):
     target_model = AutoModel.from_pretrained(
         config.model.target_model_name_or_path,
         dtype=torch.bfloat16,
-        attn_implementation="sdpa",
+        attn_implementation="eager",
```
Comment (huangxinV587) [inline scripts/data/prepare_target_cache.py]: GPU 用户也被强制走 eager?

File: scripts/data/prepare_target_cache.py [hunk 3]
```diff
@@ -397,6 +418,8 @@ def main(local_rank: int):
 
 if __name__ == "__main__":
     if os.path.exists(".git"):
-        print(f"git status:", "\n\n".join(get_git_sha(detail_info=True)))
-        print("git diff:", get_git_diff())
-    torch.multiprocessing.spawn(main, nprocs=torch.cuda.device_count())
+        try:
+            print(f"git sha:", get_git_sha())
```
Comment (huangxinV587) [inline scripts/data/prepare_target_cache.py]: debug信息与原实现不对齐

File: scripts/data/prepare_target_cache.py [hunk 4]
```diff
@@ -46,6 +46,27 @@
 torch.set_float32_matmul_precision("high")
 
 
+def _is_npu_available():
+    try:
```
Comment (huangxinV587) [inline scripts/data/prepare_target_cache.py]: 这重复的有些多...

File: deepspec/modeling/dspark/common.py [hunk 5]
```diff
@@ -96,14 +97,42 @@ def dspark_mask_mod(b, h, q_idx, kv_idx):
...
+            device=device,
+        )
+    # Eager / non-flex path: materialize a 4D additive mask (bool -> 0/-inf).
+    # attn pattern: (mask_context | mask_draft) & is_valid_block
+    q_idx = torch.arange(q_len, device=device).view(1, 1, q_len, 1)
+    kv_idx = torch.arange(kv_len, device=device).view(1, 1, 1, kv_len)
+    q_block_id = q_idx // block_size  # (1,1,q_len,1)
+    # gather anchor_pos per q_block_id: (bsz,1,q_len,1)
+    anchor_pos_q = torch.gather(
+        anchor_positions, dim=1, index=q_block_id.view(bsz, -1)
```
Comment (huangxinV587) [inline deepspec/modeling/dspark/common.py]: 注意这里 bsz>1 的情况，前面q_idx (1, 1, q_len, 1)

**Review by rajpratham1 — APPROVED** [review]
Nice work abstracting accelerator-specific functionality into a dedicated device layer instead of spreading CUDA/NPU conditionals throughout the codebase. The changes are consistent across distributed training, checkpointing, dataloading, and runtime initialization, which should make future hardware support much easier to maintain. Before merging, I'd recommend validating checkpoint compatibility across devices, ensuring no remaining direct torch.cuda usage exists in execution paths, benchmarking the dense attention fallback on NPU, and documenting the new DEEPSPEC_DEVICE workflow for users.

File: eval.py (hunk 1 above)
Comment (sunny-infra) [reply eval.py]: Revised

File: scripts/data/prepare_target_cache.py (hunk 2 above)
Comment (sunny-infra) [reply scripts/data/prepare_target_cache.py]: Revised

File: scripts/data/prepare_target_cache.py (hunk 3 above)
Comment (sunny-infra) [reply scripts/data/prepare_target_cache.py]: Revised

File: scripts/data/prepare_target_cache.py (hunk 4 above)
Comment (sunny-infra) [reply scripts/data/prepare_target_cache.py]: New Centralized Hardware Abstraction Layer: deepspec/utils/device.py

File: deepspec/modeling/dspark/common.py (hunk 5 above)
Comment (sunny-infra) [reply deepspec/modeling/dspark/common.py]: Revised
